pynumaflow/function/async_server.py

```diff
@@ -0,0 +1,315 @@
+UDFReduceCallable = Callable[[List[str], AsyncIterable[Datum], Metadata], Messages]
+_PROCESS_COUNT = multiprocessing.cpu_count()
+MAX_THREADS = int(os.getenv("MAX_THREADS", 0)) or (_PROCESS_COUNT * 4)
+
+
+async def datum_generator(
+    request_iterator: AsyncIterable[udfunction_pb2.DatumRequest],
+) -> AsyncIterable[Datum]:
+    async for d in request_iterator:
+        datum = Datum(
+            keys=list(d.keys),
+            value=d.value,
+            event_time=d.event_time.event_time.ToDatetime(),
+            watermark=d.watermark.watermark.ToDatetime(),
+            metadata=DatumMetadata(
+                msg_id=d.metadata.id,
+                num_delivered=d.metadata.num_delivered,
+            ),
+        )
+        yield datum
+
+
+class AsyncServer(udfunction_pb2_grpc.UserDefinedFunctionServicer):
+    """
+    Provides an interface to write a User Defined Function (UDFunction)
+    which will be exposed over gRPC.
+
+    Args:
+        map_handler: Function callable following the type signature of UDFMapCallable
+        mapt_handler: Function callable following the type signature of UDFMapTCallable
+        reduce_handler: Function callable following the type signature of UDFReduceCallable
+        sock_path: Path to the UNIX Domain Socket
+        max_message_size: The max message size in bytes the server can receive and send
+        max_threads: The max number of threads to be spawned;
+                     defaults to number of processors x4
+
+    Example invocation:
+    >>> from typing import Iterator
+    >>> from pynumaflow.function import Messages, Message, MessageTs, MessageT, \
+    ...     Datum, Metadata, AsyncServer
+    ... import aiorun
+    ...
+    >>> async def map_handler(key: [str], datum: Datum) -> Messages:
+    ...   val = datum.value
+    ...   _ = datum.event_time
+    ...   _ = datum.watermark
+    ...   messages = Messages(Message(val, keys=keys))
+    ...   return messages
+    ...
+    ...
+    >>> async def reduce_handler(key: str, datums: Iterator[Datum], md: Metadata) -> Messages:
+    ...   interval_window = md.interval_window
+    ...   counter = 0
+    ...   async for _ in datums:
+    ...     counter += 1
+    ...   msg = (
+    ...       f"counter:{counter} interval_window_start:{interval_window.start} "
+    ...       f"interval_window_end:{interval_window.end}"
+    ...   )
+    ...   return Messages(Message(value=str.encode(msg), keys=keys))
+    ...
+    >>> grpc_server = AsyncServer(
+    ...   reduce_handler=reduce_handler,
+    ...   map_handler=map_handler,
+    ... )
+    >>> aiorun.run(grpc_server.start())
+    """
+
+    def __init__(
+        self,
+        map_handler: UDFMapCallable = None,
+        mapt_handler: UDFMapTCallable = None,
+        reduce_handler: UDFReduceCallable = None,
+        sock_path=FUNCTION_SOCK_PATH,
+        max_message_size=MAX_MESSAGE_SIZE,
+        max_threads=MAX_THREADS,
+    ):
+        if not (map_handler or mapt_handler or reduce_handler):
```

Comment:
we may want to remove mapt_handler if we are not planning on support async for mapt

Reply:
i mean the check

Reply:
Should we remove similarly for the other two servers as well in terms of current support?